### PR TITLE
Build app

### PR DESCRIPTION
**Pull request by BenMKT**
In this PR , I add the following changes to build a web application where you can manage your budget

- A simple welocme page with the name of the app , and links to the sign up and log in pages.
- Use devise for authentication and Allow the user to register in the app with full name, email and passwordT and the user can log into the app using email and password.
- If the user is not logged in, they can't access pages that require the user to be logged in
### Categories

-When the user logs in, they are presented with the categories page.
For each category, the user can see their name, icon and the total amount of all the transactions that belong to that category.

- For each category, the user is able to delete it
- When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
- There is a button "add a new category" at the bottom that brings the user to the page to create a new category. Here the user is able to select the icon name and the system gives the corresponding icon for it.
- The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
- When the user clicks on the "Back" button (<), the user navigates to the home page.

### Transactions page

- For a given category, the list of transactions is presented, ordered by the most recent.
- At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
- When the user clicks on the "Back" button (<), the user navigates to the home page.
- There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
- The user fills out a form to create a new transaction with:
- The user clicks the "Save" button to create the new transaction and is taken to the transactions page for that category.
- When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.

### Testing

- Unit and Integration Tests

### CODE REVIEW

File: app/views/transactions/new.html.erb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    </div>
+    <div class="field">
+      <%= f.label :category %>
+      <%= f.collection_select :category_id, @categories, :id, :name, {}, { required: true, class: "transaction-input" } %>
+    </div>
+    <%= f.submit 'Save', class: "submit-button" %>
+  <% end %>
+
+  <%= link_to 'Back', category_transactions_path(@category), class: "back-link" %>
+</div>
```
Comment (thecodechaser) [inline app/views/transactions/new.html.erb:21]: - [x] On "Add a new transaction" page after filling all details when user clicks on the save button the user should redirected to transaction page for that selected category(from dropdown) and transaction should be created in the selected category as requested in the project requirements. Currently transaction is created in the same category from where user opens the "Add a new transaction" page. Please make sure to fix this issue.

File: app/views/categories/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            </div>
+          </div>
+        </div>
+      </li>
+    <% end %>
+  </ul>
+</div>
+<div class="new-category-button-container">
+  <%= link_to 'Add a new category', new_category_path, class: 'new-category-button' %>
+</div>
```
Comment (thecodechaser) [inline app/views/categories/index.html.erb:134]: - [x] Please make sure the buttons in the application matches with the given design.

### Your design

image

image

### Given design

image

File: app/views/transactions/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+          Transaction Date: <%= transaction.created_at %>
+        </div>
+      </li>
+    <% end %>
+  </ul>
+  </div>
+  <div class="new-category-button-container">
+  <%= link_to 'Add a new transaction', new_category_transaction_path(@category), class: 'new-category-button' %>
+</div>
+  </div>
```
Comment (thecodechaser) [inline app/views/transactions/index.html.erb]: - [x] On "Transaction page" first time when user add any transaction the total amount shows at top but if user comes again to that transaction page the total amount is showing at the bottom. Please make sure to fix this issue, total amount should be displayed at top as requested in the projqct requirements.
image

image

**Review by thecodechaser — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ 

## Hi @BenMKT👋, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

image

## To Highlight 🎉
- All tests are passed. ✔️
- Nice UI design. ✔️
- No linter errors. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- Gitflow is followed. :heavy_check_mark:
- PR has a good title and summary. ✔️



## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/transactions/new.html.erb (hunk 1 above)
Comment (BenMKT) [reply app/views/transactions/new.html.erb:21]: - Modified the transactions controller action that handles the creation of the transaction to resolve bug.

File: app/views/categories/index.html.erb (hunk 2 above)
Comment (BenMKT) [reply app/views/categories/index.html.erb:134]: - Refactor application CSS to resolve bug.

File: app/views/transactions/index.html.erb (hunk 3 above)
Comment (BenMKT) [reply app/views/transactions/index.html.erb]: - Refactor application CSS classes to resolve bugs.

File: app/views/categories/index.html.erb (hunk 2 above)
Comment (Munsa1) [inline app/views/categories/index.html.erb:134]: - [x] As requested by the previous reviewer in this [comment](https://github.com/BenMKT/Budget-App/pull/1#discussion_r1403266016) kindly fix the button design to match the given design

Your design
image

image


Given design
image


- First, let's use the given color swatch for the buttons
image

- Second, add the white border to your buttons as is in the given design (the first button has a border kindly do so for the second one as well. After creating a transaction for the first time the button doesn't show well. It only shows better after reloading the page. Check the video below)

[cinnamon-2023-11-23T213710+0200.webm](https://github.com/BenMKT/Budget-App/assets/81932478/c7a3bca1-308c-4521-922b-71c1a386b6cf)



- Third, your button should span a width similar to the given design (In the given design the button fills up almost the whole width of the mobile design)

**Review by Munsa1 — CHANGES_REQUESTED** [review]
Hi @BenMKT, 

Great work on making the changes 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


## Suggested changes
_Check the comments under the review._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

File: app/views/categories/index.html.erb (hunk 2 above)
Comment (BenMKT) [reply app/views/categories/index.html.erb:134]: - Added a `a.new-category-button` class with styling to resolve bug.

File: app/assets/stylesheets/application.css [hunk 4]
```diff
@@ -1,15 +1,415 @@
...
+a.new-category-button:hover {
+  background-color: #4a8e18;
+}
+
+div.new-category-button-container {
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  background-color: #5fb523;
+}
```
Comment (KanzaTahreem) [inline app/assets/stylesheets/application.css]: - [x] As noted by the previous reviewer [here](https://github.com/BenMKT/Budget-App/pull/1#:~:text=After%20creating%20a%20transaction%20for%20the%20first%20time%20the%20button%20doesn%27t%20show%20well.%20It%20only%20shows%20better%20after%20reloading%20the%20page.%20Check%20the%20video%20below), the button fails to retain the intended style after the creation of a transaction 🐛However, it displays correctly in accordance with the design specifications when redirected from the home page.
- [x] As mentioned earlier, please ensure that the button spans a width similar to the provided design⛏🤓 _Please have a look at the attached video 📹_

https://github.com/BenMKT/Budget-App/assets/106119105/109278e6-6eae-4cb4-9c53-d5a530171bc6

**Review by KanzaTahreem — CHANGES_REQUESTED** [review]
# Changes Required 🔁

Hi @BenMKT 

Great work on making the changes requested by a previous reviewer 👏! I really appreciate the effort you put in!  **_However, there are few things that still need to be fixed. But you are almost there!_** 💪 🏁

so close


## Highlights 🎯

There are some excellent points that you have made perfectly. Keep up the great work! 👏
- ✅ Good job with the design
- ✅ Great job with CSS
- ✅ Descriptive PR and README file
- ✅ Linters are working properly

## Required Changes 🛑

_Please check the comments under the review._

## Optional suggestions ❕

N/A

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👨‍💻💻

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/assets/stylesheets/application.css (hunk 4 above)
Comment (BenMKT) [reply app/assets/stylesheets/application.css]: - Added `padding` and `height` to `div.new-category-button-container` to resolve bug.

Comment (peterrobert) [reply app/assets/stylesheets/application.css]: This change has been done well. The button now spans the entire width

**Review by peterrobert — APPROVED** [review]
Hi @BenMKT, 


### Highlights!

[ ] Good feature branch name ✔️
[ ] Descriptive pull request ✔️
[ ] Great work making the changes requested by the previous reviewer ✔️

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread or tag me @peterrobert if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have only 3 code reviews per project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**